ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh

```diff
@@ -445,31 +449,74 @@ EOF
   done
 }
 
-# inject_spot_instance_config is an AWS specific option that enables the use of AWS spot instances for worker nodes
+# inject_spot_instance_config is an AWS specific option that enables the
+# use of AWS spot instances.
+# PARAMS:
+# $1: Path to base output directory of `openshift-install create manifests`
+# $2: Either "workers" or "masters" to enable spot instances on the
+#     compute or control machines, respectively.
 function inject_spot_instance_config() {
   local dir=${1}
+  local mtype=${2}
 
-  if [ ! -f /tmp/yq ]; then
-    curl -L "https://github.com/mikefarah/yq/releases/download/3.3.0/yq_linux_$( get_arch )" \
-    -o /tmp/yq && chmod +x /tmp/yq
-  fi
 
-  PATCH="${SHARED_DIR}/machinesets-spot-instances.yaml.patch"
-  cat > "${PATCH}" << EOF
-spec:
-  template:
-    spec:
-      providerSpec:
-        value:
-          spotMarketOptions: {}
-EOF
+  get_yq
 
-  for MACHINESET in $dir/openshift/99_openshift-cluster-api_worker-machineset-*.yaml; do
-    /tmp/yq m -x -i "${MACHINESET}" "${PATCH}"
-    echo "Patched spotMarketOptions into ${MACHINESET}"
+  # Find manifest files
+  local manifests=
+  case "${mtype}" in
+    masters)
+      # Spot masters only works for CAPI today. Validate.
+      if ! [[ "$(/tmp/yq r "${SHARED_DIR}/install-config.yaml" 'featureGates[*]')" == *'ClusterAPIInstall=true'* ]]; then
+        echo "Spot instances for masters can only be used with CAPI installs."
+        exit 1
+      fi
+      manifests="${dir}/cluster-api/machines/10_inframachine_*.yaml"
+      ;;
+    workers)
+      # Pick up both -machines- and -machinesets-
+      manifests="${dir}/openshift/99_openshift-cluster-api_*-machine*.yaml"
+      ;;
+    *)
+      echo "Invalid machine type '$mtype' passed to inject_spot_instance_config; expected 'masters' or 'workers'"
+      exit 1
+      ;;
+  esac
+
+  # Inject spotMarketOptions into the appropriate manifests
+  local prefix=
+  local found=false
+  # Don't rely on file names; iterate through all the manifests and match
+  # by kind.
+  for manifest in $manifests; do
+    kind=$(/tmp/yq r "${manifest}" kind)
+    case "${kind}" in
+      MachineSet)
+          [[ "${mtype}" == "workers" ]] || continue
+          prefix='spec.template.spec.providerSpec.value'
+          ;;
+      Machine)
+          [[ "${mtype}" == "workers" ]] || continue
+          prefix='spec.providerSpec.value'
+          ;;
+      AWSMachine)
+          prefix='spec'
+          ;;
+      *)
```

Comment:
Mm, I left it out because it doesn't work yet. Also, I feel like we would want _some_ kind of check or feedback when `SPOT_MASTERS` is requested for a configuration that doesn't have your [enablement](https://github.com/openshift/installer/pull/8349), vs that job silently not using spot masters. (Open to suggestions as to how to do that btw.)

Reply:
Don't we need to edit the CPMS for the CAPA case as well? Otherwise there will be a config drift, and the MCO will recreate the masters with no spot instances (iiuc).

Reply:
Mm, that makes sense, will fix.

Reply:
Done.